File: web/src/app/service/product.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {AuthenticationService} from './authentication.service';
import {Observable} from 'rxjs';
import {AppComponent} from '../app.component';
import {Product} from '../model/product';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  constructor(private http: HttpClient,
              private authenticationService: AuthenticationService) { }

  getAll(): Observable<any> {
    return this.http.get(AppComponent.API_URL + '/products', this.authenticationService.jwt());
  }

  create(product: Product): Observable<any> {
    return this.http.post(AppComponent.API_URL + '/products', product, this.authenticationService.jwt());
  }

  getById(id): Observable<any> {
    return this.http.get(AppComponent.API_URL + '/products/' + id, this.authenticationService.jwt());
  }

  update(product: Product): Observable<any> {
    return this.http.put(AppComponent.API_URL + '/products', product, this.authenticationService.jwt());
  }

}

File: src/main/java/com/nc/webapp/service/ProductService.java
package com.nc.webapp.service;

import com.nc.webapp.dto.ProductSalesCountDTO;
import com.nc.webapp.model.Product;

import java.util.List;

public interface ProductService {
    Product createOrUpdate(Product product);
    List<Product> getAll();
    Product getById(Integer id);
    List<ProductSalesCountDTO> getProductSalesCountReportData();
}
